AzureMonitor: Move DefaultSubscriptions field to separate component

aangelisc opened the pull request:
Decouples the `DefaultSubscription` component from the `AzureCredentialsForm`. This component is not currently needed to generalise `AzureAuth` as it is specific to Azure Monitor.

Fixes #60780

aangelisc committed:
Move DefaultSubscriptions field to separate component

- Add separate component
- Update CredentialsForm
- Update tests
- Remove redundant test

kostrse commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MonitorConfig.tsx:
```diff
@@ -38,7 +46,16 @@ export const MonitorConfig: FunctionComponent<Props> = (props: Props) => {
+      >
+        <DefaultSubscription
+          subscriptions={subscriptions}
+          credentials={credentials}
```
Could you please completely decouple the subscription from credentials?

Default subscription is just a filed `subscriptionId` in JsonData. We need to remove it from `AzureCredentialsBase` ([here](https://github.com/grafana/grafana/blob/e8c2091537a0641d71c3790ec88b5ae34da57bc6/public/app/plugins/datasource/grafana-azure-monitor-datasource/types/types.ts#L38)), to make the interface look same as it defined in ADX ([here](https://github.com/grafana/azure-data-explorer-datasource/blob/42176219a831d41fc1e1999617e4686d43ca8b84/src/components/ConfigEditor/AzureCredentials.ts#L19)).

andresmgot replied:
I agree this makes sense. It could be a breaking change so some migration code may be needed.

kostrse replied:
Data schema should not change. The field `subscriptionId` stays unchanged. `AzureCredentialsBase` is in-memory only, so changing it should not require any migration.

aangelisc replied:
I've looked into this for the past few Grafana versions and this should not be a breaking change 😊

aangelisc committed:
Remove defaultSubscriptionId

- Update MonitorConfig to appropriately set subscriptionId
- Update DefaultSubscription component to make use of JsonData
- Remove references to defaultSubscriptionId
- Update tests
- Update mockInstanceSettings for flexibility

aangelisc committed:
Merge branch 'main' into credentials-form-update

andresmgot reviewed (approved):
LGTM, thanks for checking!

kostrse commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/components/DefaultSubscription.tsx:
```diff
@@ -0,0 +1,109 @@
+
+interface Props {
+  options: AzureDataSourceJsonData;
+  credentials: AzureCredentials;
```
`DefaultSubscripton` doesn't actually need to know anything about credentials.

In my opinion, it could be implemented with only three main properties:
- `subscriptions` - the list of subscriptions to select from (supplied from outside and empty if not yet loaded)
- `subscriptionId`\`selectedSubscription` - the subscription which was actually selected, `undefined` if no subscription has been selected.
- `onSubscriptionChange` - callback when another subscription being selected, so the outside code, `MonitorConfig` can apply it to datasource settings.

Every time `AzureCredentialsForm` changes credentials, it calls` onCredentialsChange` and `MonitorConfig` clears the `subscriptions` state variable which would propagate to the `DefaultSubscription` control.

`MonitorConfig` could also be responsible for the asynchronous call to fetch the list of subscriptions since it knows when credentials change and knows when to re-fetch the subscriptions.

kostrse reviewed (changes requested):
Let's completely break the dependency between default subscription and credentials.

`DefaultSubscription` control doesn't need to konw about `credentials` and `subscriptionId` can be set by `MonitorConfig` directory to JsonData using `updateOptions` without having to use `updateCredentials` for this purpose.

aangelisc commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/components/AzureCredentialsForm.test.tsx:
```diff
@@ -55,15 +58,26 @@ describe('Render', () => {
-      },
+      credentials,
+      children: (
+        <DefaultSubscription
```
I can separate this component from the test 👍🏽

aangelisc commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/credentials.ts:
```diff
@@ -109,14 +108,14 @@ export function getCredentials(options: AzureDataSourceSettings): AzureCredentia
   options: AzureDataSourceSettings,
-  credentials: AzureCredentials
+  credentials: AzureCredentials,
+  subscriptionId?: string
```
I've updated this function to remove the `subscriptionId` dependency.

aangelisc committed:
Review

- Add DefaultSubscription test and remove component from AzureCredentialsForm test
- Remove onCredentialsChange from DefaultSubscription
- Set subscription as a part of onSubscriptionChange

aangelisc commented:
I've made some changes based on your comments @kostrse.

For now, I'd still like `DefaultSubscription` to have `credentials` as a prop to keep the logic for identifying if the credentials are complete within this component as it's the only component that makes use of this right now. This also doesn't affect moving the `AzureCredentialsForm` component to the SDK. 

Also, I don't think there's any reason for `AzureCredentialsForm` to not have the `DefaultSubscription` component as a child. There are ongoing discussions about how the UI/UX of config pages should function and I'd rather not have a component that appears isolated on the page at the moment.

aangelisc committed:
Merge branch 'main' into credentials-form-update
